File: backend/src/main/java/nlu/fit/backend/service/CartService.java
package nlu.fit.backend.service;

import nlu.fit.backend.dto.cart.AddToCartRequest;
import nlu.fit.backend.dto.cart.CartSummaryResponseDto;

public interface CartService {
    CartSummaryResponseDto getCartSummary(String email);
    
    CartSummaryResponseDto addItemToCart(String email, AddToCartRequest request);
    
    CartSummaryResponseDto updateItemQuantity(Long cartItemId, int quantity, String email);
    
    CartSummaryResponseDto removeItem(Long cartItemId, String email);
    
    CartSummaryResponseDto applyPromoCode(String promoCode, String email);
    
    void clearCart(Long userId);
}

File: frontend/src/modules/cart/service/cartService.ts
import { apiCall } from '../../../api/client';
import { ENDPOINTS } from '../../../api/constants';
import type { CartItem, CartSummaryResponse } from '../model';

export async function getCart(): Promise<CartItem[]> {
  return apiCall<CartItem[]>(`${ENDPOINTS.CART}`);
}

// Add to cart (productId, optional size, quantity)
export async function addToCart(item: { productId: string; size?: number; quantity?: number }): Promise<void> {
  await apiCall<void>(`${ENDPOINTS.CART}/add`, {
    method: 'POST',
    body: JSON.stringify(item),
  });
}

// Update quantity of an existing cart item
export async function updateItemQuantity(cartItemId: number, quantity: number): Promise<void> {
  await apiCall<void>(`${ENDPOINTS.CART}/items/${cartItemId}`, {
    method: 'PUT',
    body: JSON.stringify({ quantity }),
  });
}

// Remove a cart item
export async function removeItem(cartItemId: number): Promise<void> {
  await apiCall<void>(`${ENDPOINTS.CART}/items/${cartItemId}`, {
    method: 'DELETE',
  });
}

// Get full cart summary (includes totals, promo, etc.)
export async function getCartSummary(): Promise<CartSummaryResponse> {
  return apiCall<CartSummaryResponse>(`${ENDPOINTS.CART}`);
}

// Apply promo code
export async function applyPromoCode(promoCode: string): Promise<CartSummaryResponse> {
  return apiCall<CartSummaryResponse>(`${ENDPOINTS.CART}/promo/apply`, {
    method: 'POST',
    body: JSON.stringify({ promoCode }),
  });
}

export const cartService = {
  getCart,
  addToCart,
  updateItemQuantity,
  removeItem,
  getCartSummary,
  applyPromoCode,
};
